sympy/polys/numberfields/forms.py

```diff
@@ -0,0 +1,449 @@
+        return igcd(g, self.denom) == 1
+
+    def reduced(self):
+        """Produce a reduced version of this rep."""
+        g = igcd(self.denom, *self.coeffs)
+        return StandardRep(self.T, [c // g for c in self.coeffs], self.denom // g)
+
+    def clone(self):
+        return StandardRep(self.T, self.coeffs[:], self.denom)
+
+    standard_rep = clone
+
+    def in_rationals(self):
+        return all(c == 0 for c in self.coeffs[1:])
+
+    @property
+    def is_zero(self):
+        return self == 0
+
+    @classmethod
+    def zero(cls, T):
+        return cls.from_int(T, 0)
+
+    @classmethod
+    def one(cls, T):
+        return cls.from_int(T, 1)
+
+    @classmethod
+    def from_int(cls, T, a):
+        return cls(T, [a] + [0] * (T.degree() - 1), 1)
+
+    @classmethod
+    def from_rational(cls, T, a):
+        a = sympify(a)
+        n, d = a.as_numer_denom()
+        return cls(T, [n] + [0] * (T.degree() - 1), d)
+
+    @classmethod
+    def from_poly(cls, T, f):
+        n, k = T.degree(), f.degree()
+        if k >= n:
+            f = f % T
+        if f == 0:
+            return cls.zero(T)
+        d, g = f.clear_denoms()
+        c = list(reversed(g.all_coeffs()))
+        ell = len(c)
+        z = [ZZ(0)] * (n - ell)
+        return cls(T, c + z, d)
+
+    @classmethod
+    def from_matrix_col(cls, T, M, j, d):
+        return cls(T, M.col(j).flat(), d)
+
+    def __eq__(self, other):
+        if isinstance(other, StandardRep):
+            s_r = self if self.is_reduced() else self.reduced()
+            o_r = other if other.is_reduced() else other.reduced()
+            return s_r.denom == o_r.denom and s_r.coeffs == o_r.coeffs
+        if is_rat(other):
+            other = sympify(other)
+            return self.in_rationals() and other.numerator * self.denom == other.denominator * self.coeffs[0]
+        return NotImplemented
+
+    def __add__(self, other):
+        if isinstance(other, StandardRep):
+            d, e = self.denom, other.denom
+            m = ilcm(d, e)
+            u, v = m // d, m // e
+            r = StandardRep(
+                self.T, [u * a + v * b for a, b in zip(self.coeffs, other.coeffs)], m)
+            return r.reduced()
+        if is_rat(other):
+            other = sympify(other)
+            return self + StandardRep.from_rational(self.T, other)
+        return NotImplemented
+
+    __radd__ = __add__
+
+    def __neg__(self):
+        return self * (-1)
+
+    def __sub__(self, other):
+        return self + (-other)
+
+    def __rsub__(self, other):
+        return -self + other
+
+    def __mul__(self, r):
+        if isinstance(r, StandardRep):
+            f, g = self.poly(), r.poly()
+            e, h = (f * g % self.T).clear_denoms()
+            return StandardRep.from_poly(self.T, h) // e
+        if is_rat(r):
+            r = sympify(r)
+            a, b = r.as_numer_denom()
+            s = StandardRep(self.T, [a*c for c in self.coeffs], b*self.denom)
```

Comment:
If this is supposed to use the ZZ domain then `sympify` should not be used as it gives `Integer` rather than `MPZ`. SymPy `Expr` types are infectious: if a single one creeps in it can ruin everything.

I suggest adding a check in `__init__`:
```python
assert all(ZZ.of_type(c) for c in coeffs)
```
If the check slows things down then it can be disabled later but at least the internals here need to be sorted out first.